src/chef_user.erl

```
@@ -92,124 +84,158 @@ id(#chef_user{id = Id}) ->
-  ok;
+    ok;
 valid_password(_Password) ->
-  error.
-
-
+    error.
 
+%% Returns the subset of permitted fields that are
+%% present in JSON_SERIALIZABLE
+whitelisted_values(EJ) ->
+    [{X, ej:get({X}, EJ) } || X <- ?JSON_SERIALIZABLE, ej:get({X}, EJ)  =/= undefined ].
 assemble_user_ejson(#chef_user{username = Name,
-                               public_key = PubKey,
-                               admin = Admin},
+                               public_key = KeyOrCert,
+                               email = Email,
+                               serialized_object = SerializedObject},
                     _OrgId) ->
-    {[{<<"name">>, Name},
-      {<<"public_key">>, PubKey},
-      {<<"admin">>, Admin}]}.
+    EJ = chef_json:decode(SerializedObject),
+    % public_key can mean either public key or cert.
+    % if it's a cert, we need to extract the public key -
+    % we don't want to hand the cert back on user GET.
+    RealPubKey = real_pub_key(KeyOrCert),
+    User1 = [{<<"username">>, Name},
+             {<<"public_key">>, RealPubKey},
+             {<<"email">>, Email}],
+    User2 = whitelisted_values(EJ),
+    { User1 ++ User2 }.
+
+
+real_pub_key(Data) ->
+    case chef_object_base:key_version(Data) of
+        ?KEY_VERSION ->
+            Data;
+        ?CERT_VERSION ->
+            chef_authn:extract_pem_encoded_public_key(Data)
+    end.
 
 %% @doc Convert a binary JSON string representing a Chef User into an
 %% EJson-encoded Erlang data structure.
 -spec parse_binary_json(binary()) -> {ok, ej:json_object()}. % or throw
 parse_binary_json(Bin) ->
-    parse_binary_json(Bin, create).
+  parse_binary_json(Bin, create).
 
 -spec parse_binary_json(binary(), create | update) -> {ok, ej:json_object()}. % or throw
 parse_binary_json(Bin, Operation) ->
   User = chef_object_base:delete_null_public_key(chef_json:decode(Bin)),
   %% If user is invalid, an error is thown
+  validate_user(User, user_spec(common)),
```

ej will fail fast, and that bubbles up as a 400 with the associated text.  Agreed it would be better to check them all at once, though I think that's something to add to the list.